04-onion-routing.md

```diff
@@ -261,6 +268,197 @@ The reader:
+```
+    s_1  s_2  s_3    SHARES OF ROOT SEED
+                   
+     |    |    |     
+     └──┐ | ┌──┘       R = s_1 ^ s_2 ^ s_3
+        V V V        
+                   
+          R          ROOT SEED
+                   
+        | | |        
+     ┌──┘ | └──┐       P_i = SHA256(R || child_index_i)
+     V    V    V     
+                   
+    P_1  P_2  P_3    CHILD PREIMAGES
+                   
+     |    |    |     
+     |    |    |       H_i = SHA256(P_i)
+     V    V    V     
+                   
+    H_1  H_2  H_3    CHILD HASHES
+```
+
+### Rationale
+
+The inclusion of a `payment_id` in `option_amp` allows the receiver to map an
+incoming AMP payment to a particular invoice. The sender should set the
+`payment_id` to the `payment_hash` of the invoice they are trying to pay,
+permitting the receiver to enforce custom parameters, e.g. CLTV deltas,  and
+unify tracking of AMP payments with the existing invoicing system.
+
+At the same time, AMP payments can be made spontaneously (without and invoice),
+since the sender generates all of the necessary secrets. To do so, the sender
+leaves the `payment_id` blank, which can be used to facilitate secure donations.
+
+In the event that two payments are made with the same `payment_id`, either to
+the same invoice for both are spontaneous, a second identifier is introduced
+called the `stream_id`. The `stream_id` should be unique to each HTLC set sent
+by the sender, and allows the receiver to distinguish concurrent payments that
+collide on `payment_id`.
+
+Both the `stream_id` and `payment_id` are only known to sender and receiver,
+preventing intermediaries from introducing griefing via colliding payment
+identifiers with high probability.
+
+The `child_index` is included in the final payload so that the receiver can
+gracefully tolerate reordering of the partial payments. When each `child_index`
+is unique, this offers decorrelation of the partial payments, since they bear
+different child payment hashes while traversing the network.
+
+The `total_msat` field is used to determine when all partial payments have been
+received. If the AMP is paying an invoice, this also allows the sender to
+securely overpay an invoice, for instance, if the invoice's `amount` is
+unspecified. If the AMP is spontaneous, this allows the sender to communicate
+the exact value to be received in a end-to-end authenticated manner, preventing
+certain classes of attacks where intermediaries can steal up to the overpaid
+amount.
+
+The entire payment is contingent on the receiver being able to reconstruct the
+root seed `r`, which prevents the receiver from pulling any of the partial
+payments until the entirety of the HTLC set arrives. This is enforced by the
+n-of-n secret shares provided in the final hop payload of each arriving partial
+payment.
+
+None of the requirements enforce that more than one HTLC is sent, permitting the
+base case of 1 HTLC to function as a standalone spontaneous payment.
```

Comment:
:clap: I think this is important and this is why I view this work more as a "spontaneous payments" overall feature